write-fonts/src/graph/splitting.rs

```diff
@@ -1066,4 +1063,81 @@ mod tests {
+                None => rec,
+            }
+        }
+
+        let class_def1 = make_class_def(1, 4, 1);
+        let class_def2 = make_class_def(3, 3, 1);
+        let coverage = class_def1.iter().map(|(gid, _)| gid).collect();
+        let actual_device_table = Device::new(12, 15, &[118, 119, 127, 99]);
+        // sanity check the size of the device table, these are weird:
+        assert_eq!(crate::dump_table(&actual_device_table).unwrap().len(), 10);
+        let class1_records = vec![Class1Record::new(vec![
+            Class2Record::new(
+                make_val_record(5, None),
+                make_val_record(6, DeviceOrVariationIndex::variation_index(4, 20)),
+            ),
+            Class2Record::new(
+                make_val_record(7, DeviceOrVariationIndex::Device(actual_device_table)),
+                // a duplicate table
+                make_val_record(8, DeviceOrVariationIndex::variation_index(4, 20)),
+            ),
+            Class2Record::new(
+                make_val_record(9, DeviceOrVariationIndex::variation_index(6, 9)),
+                make_val_record(10, None),
+            ),
+        ])];
+        let ppf2 = PairPos::format_2(coverage, class_def1, class_def2, class1_records);
+
+        // now we need to pretend we're in the split_pair_pos_format_2 fn
+        let mut graph = TableWriter::make_graph(&ppf2);
+        assert!(graph.pack_objects());
+        let root_id = graph.root;
+        let ppf2_data = &graph.objects[&root_id];
+        let ppf2 = ppf2_data.reparse::<rgpos::PairPosFormat2>().unwrap();
+        assert_eq!(ppf2.class1_records().len(), 1);
+        let c1rec = ppf2.class1_records().get(0).unwrap();
+        let mut visited = HashSet::new();
+        let mut next_device_offset = 3;
+        assert_eq!(c1rec.class2_records.len(), 3);
+
+        // a little helper so we don't have to have this huge fn call in each assert
+        let mut test_closure = |record: &rgpos::ValueRecord| -> usize {
+            size_of_value_record_children(
+                record,
+                &graph,
+                &ppf2_data.offsets,
+                &mut next_device_offset,
+                &mut visited,
+            )
+        };
+
+        let c2rec1 = c1rec.class2_records().get(0).unwrap();
+        assert_eq!(test_closure(c2rec1.value_record1()), 0, "no subtables");
+        assert_eq!(test_closure(c2rec1.value_record2()), 6, "one new varidx");
+        let c2rec2 = c1rec.class2_records().get(1).unwrap();
+        assert_eq!(test_closure(c2rec2.value_record1()), 10, "a device table");
+        assert_eq!(test_closure(c2rec2.value_record2()), 0, "duplicate table");
+        let c2rec3 = c1rec.class2_records().get(2).unwrap();
+        assert_eq!(test_closure(c2rec3.value_record1(),), 6, "new varidx table");
+        assert_eq!(test_closure(c2rec3.value_record2(),), 0, "a null offset");
+        assert_eq!(next_device_offset, 7, "we visited all offsets");
+    }
```

Comment:
IMHO this would be nicer as one assert that there are the values. Less noisy bindings, error-prone get(index) calls and no need to check we visited everything.

```rust
assert_eq!(vec![
  (0, 6),  // no subtables, one new var idx
  (10, 0),// a device table with a duplicate table
  (6, 0),  // a new varidx table with a null offset
],
c1rec.class2_records().iter().map(|r| (r.value_record1(), r.value_record2()).collect::<Vec<_>>()
);
```

Reply:
huh interesting, I find personally find that slightly less clear? I prefer fine-grained asserts because it makes explicit exactly what has failed.

The check for the value of `next_device_offset` is important, because this we want to ensure that this value is incremented correctly, e.g. it is incremented for each non-null offset in each value record. we visit.

Reply:
The single assert tells you when lots of values were wrong, when you have the wrong number of values, and makes it much harder to `get(the wrong value)`. I would argue it's equivalent in terms of telling you what failed while giving greater context (were ALL the values borked? just one?) but I'll grant you it's stylistic so you could view this one as optional.